eliminate intrinsics

vtjnash opened the pull request:
After this, only 1 intrinsics (llvmcall ~~and some ccall fixes~~) remain to go to be fully interpretable and statically compilable!
- removes special parsing rules for `ccall` (and thus removes `Expr(:ccall)`), fixes #18687
- makes `Expr(:call, :ccall)` special in lowering (effectively, reserving `ccall` – as the keyword it already is) and which expands to `Expr(:foreigncall)` representing the lowered ffi call (instead of expanding to `Expr(:call, Core.Intrinsics.ccall)`, even though this couldn't really be a first-class function)
- adds error checking to all Core.Intrinsic functions
- adds support to `ccall` for calling llvm intrinsics directly (via "llvmcall" calling convention)
- the `unbox` intrinsic is now just a deprecated alias for the `box` intrinsic (aka `reinterpret`)

The ccall intrinsic now evaluates its global arguments during method definition time, eliminating the call back into the runtime from inside codegen (yay!), and making its semantics independent of runtime values (so that its behavior is statically compilable).

edit: see https://github.com/vtjnash/julep/wiki/0001:-Enhanced-static-compilation-and-C-interface

vtjnash committed:
add llvmcall calling convention to ccall

vtjnash committed:
remove magic ccall handling from parsing

vtjnash committed:
refactor ccall code into less interleaved functionality

vtjnash committed:
some cleanup of runtime-intrinsics entry points

vtjnash committed:
entirely hookup runtime-intrinsics

all intrinsics (but one) are now fully error-checking and runtime-interpreter-compatible

Sacha0 commented on base/floatfuncs.jl:
```diff
@@ -2,40 +2,40 @@
-copysign(x::Float64, y::Float64) = box(Float64,copysign_float(unbox(Float64,x),unbox(Float64,y)))
-copysign(x::Float32, y::Float32) = box(Float32,copysign_float(unbox(Float32,x),unbox(Float32,y)))
+copysign(x::Float64,  y::Float64) = box(Float64, copysign_float(x, y))
+copysign(x::Float32,  y::Float32) = box(Float32, copysign_float(x, y))
```
Extra space preceding `y` in the method signature?

Sacha0 commented on base/floatfuncs.jl:
```diff
@@ -2,40 +2,40 @@
-flipsign(x::Float64, y::Float64) = box(Float64,xor_int(unbox(Float64,x),and_int(unbox(Float64,y),0x8000000000000000)))
-flipsign(x::Float32, y::Float32) = box(Float32,xor_int(unbox(Float32,x),and_int(unbox(Float32,y),0x80000000)))
+flipsign(x::Float64,  y::Float64) = box(Float64, xor_int(box(UInt64, x), and_int(box(UInt64, y), 0x8000000000000000)))
+flipsign(x::Float32,  y::Float32) = box(Float32, xor_int(box(UInt32, x), and_int(box(UInt32, y), 0x80000000)))
```
Extra space preceding `y` in the method signature here as well?

maleadt commented:
Cool, the `llvmcall` calling convention will make some of our `llvmcall(...)` usage in CUDAnative redundant. What about function attributes though? Most NVPTX intrinsics are `readnone nounwind`.

For PTX we also need to call into `libdevice`, an IR blob providing implementations of common math functions. This blob is linked in during JIT, so regular `ccall` doesn't work (disregarding the fact that there's no `dlopen`, etc). Bottom line, it would be handy if the new `llvmcall` cc wouldn't check for isIntrinsic, even though it makes sense to require that for every other sane use case...

tkelman commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -1003,6 +1003,20 @@ For more information, please see the `LLVM Language Reference`_.
+
+.. _CUDA: http://llvm.org/docs/NVPTXUsage.html
+
+As with any ccall, it is essential to get the argument signature exactly correct.
```
code highlight `ccall`

tkelman commented on src/ccall.cpp:
```diff
@@ -1289,35 +1306,108 @@ static std::string generate_func_sig(
+        jl_error(msg.c_str()); // TODO: this leaks memory
+    }
+    if (jl_is_cpointer_type(rt) && jl_is_typevar(jl_tparam0(rt)))
+        jl_error("ccall: return type Ptr should have an element type, not Ptr{_<:T}");
```
this `Ptr{_<:T}` will be pretty confusing

vtjnash replied:
this isn't a new change

tkelman commented on src/intrinsics.cpp:
```diff
@@ -70,85 +106,80 @@ extern "C" JL_DLLEXPORT int8_t jl_is_memdebug() {
 {
+    if (t->isFloatingPointTy())
+        return t;
+    unsigned nb = (t->isPointerTy() ? t->getPrimitiveSizeInBits() : sizeof(void*));
```
is this backwards?

tkelman replied:
needs-test then if it was passing with a mistake here

tkelman commented on src/intrinsics.cpp:
```diff
@@ -70,85 +106,80 @@ extern "C" JL_DLLEXPORT int8_t jl_is_memdebug() {
+    if (t == T_float16)
+        return T_int16;
+    //if (t == T_float128)
+    //    return T_int128;
```
leave a comment or a cross-reference about why this is commented out (same below in `JL_JL_UINTT`)

vtjnash replied:
this isn't a new change

tkelman replied:
that's not a reason to leave unexplained things or lousy error messages in the code. while it's being rearranged is the right time to make these little improvements

tkelman commented on src/intrinsics.cpp:
```diff
@@ -516,191 +516,141 @@ static jl_cgval_t generic_box(jl_value_t *targ, jl_value_t *x, jl_codectx_t *ctx
-#else
-    Value *ret = builder.CreateCall3(prepare_call(jlpref_func), boxed(parg, ctx), iarg, alignarg);
+#ifdef JL_NEED_FLOATTEMP_VAR
+        // Target platform might carry extra precision.
```
excess indent

tkelman commented:
needs a test for #18687

vtjnash commented:
this isn't that kind of fix. this "fixes" the issue by deleting the special parsing code and the tests for it (of which apparently there were none).

tkelman commented:
doesn't matter how it's being fixed. add a test.

vtjnash commented:
It's not fixed, it's deleted, along with all tests for it.

tkelman commented:
That issue is about the fact that it works differently on master. To close that issue, and just in general, should add tests that it no longer works differently in this pr. Not asking for much here, tests and comments. Why are you fighting that?

vtjnash commented:
I'm not fixing that issue, I'm deleting the feature that the referenced issue complained about.

tkelman reviewed (dismissed):
So what? You're replacing it with a new implementation that **needs to be tested** that it no longer has the undesirable behavior.

vtjnash commented:
> So what? You're replacing it with a new implementation that needs to be tested that it no longer has the undesirable behavior.

No, it's completely gone. The parser no longer has any knowledge of `ccall`.

tkelman commented:
The implementation now happening in lowering instead of parsing doesn't change the fact that the parser behavior change for ccall should be tested and verified.

vtjnash commented:
How does the parser not implementing ccall not eliminate the need to test how the parser implements ccall? The parser no longer has *any* special behavior related to the symbol `ccall`, it is handled by exactly the same code as handles all `Expr(:call)` form expressions.

tkelman commented:
Because the parser used to implement ccall. If it hadn't had a custom implementation before, it would be fine. This is a change in behavior specifically for ccall, and that's the part that should now be tested.

tkelman commented:
why is llvmcall2 a separate file here? would the "nonsensical valid conversions and errors" tests have caught the mistake from https://github.com/JuliaLang/julia/pull/18754#discussion_r81501204 ?

vtjnash commented:
It tests v2 of llvmcall. Yes, it includes coverage of that case.

JeffBezanson commented:
It occurs to me that after this change we should add t-functions for all intrinsics (previously we relied on their results always being passed to `box`).

vtjnash commented:
Yes, that's also on my list of eventually fixes for here (along with completely removing unbox and removing most calls to box)

vtjnash committed:
eliminate ccall intrinsic

vtjnash committed:
eliminate uniqueness requirement on T_jlvalue (now rendered simply as T_pint8)

vtjnash commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`. This is also a good candidate for PkgEval before merging (since it changes the behavior of ccall)

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/18754#issuecomment-253839513):
```
NanosoldierError: failed to run benchmarks against primary commit: ErrorException("failed process: Process(`sudo cset shield -e su nanosoldier -- -c ./benchscript.sh`, ProcessExited(1)) [1]")
```Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/b4315662641f569288b090a8c7d6fa8247a05841/92bd0e0_vs_4ba21aa/report.md)
cc @jrevels

vtjnash committed:
add back syntax-deprecation warning for macros that were manually expanding to Expr(:ccall)

vtjnash commented:
Let's try that again, with the deprecation in place. @nanosoldier `runbenchmarks(ALL, vs=":master")`.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/18754#issuecomment-253862276) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/61182cf6c2ff31b8db4258acf548dabccd2794bb/43cc839_vs_51e9247/report.md). cc @jrevels

tkelman commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -51,24 +51,28 @@ function, all inputs must be passed by reference.
+
+   OR
+
+   a function pointer (for example, from :func:``dlsym``).
```
`:func` references use single backticks, I believe

vtjnash replied:
it seemed like `make doc` didn't care when I tested both ways

tkelman replied:
did it render correctly? there are zero instances of ```git grep -n 'func:``'``` currently

vtjnash replied:
upon rechecking, no it doesn't

tkelman commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -533,30 +539,40 @@ the field that will have the greatest size (potentially including padding).
+For example, ``f{T}(x::T) = ccall(:valid, Ptr{T}, (Ptr{T},), x)``
+is valid, since ``Ptr`` is always a word-size bitstype.
+But, ``g{T}(x::T) = ccall(:notvalid, T, (T,), x)``
+is not valid, since the type layout of T is not known statically.
```
code highlight T

tkelman commented on src/codegen.cpp:
```diff
@@ -292,8 +291,15 @@ int32_t jl_jlcall_api(const void *function)
+    const Function *F = cast<const Function>((const Value*)function);
+    StringRef Name = F->getName();
+    if (Name.startswith("japi3_"))
+       return 3;
```
indent not quite right

vtjnash commented on src/intrinsics.cpp:
```diff
@@ -38,9 +39,43 @@ static void jl_init_intrinsic_functions_codegen(Module *m)
+    float_func[fpiseq] = true;
+    float_func[fpislt] = true;
+    float_func[abs_float] = true;
+    //float_func[copysign_float] = false;
```
so that you can see it wasn't forgotten in the list

tkelman replied:
why is it false then?

tkelman replied:
could you say that in a comment then?

tkelman commented:
`Ptr{_<:T}` is still really confusing in one of the `ccall` error messages, and the commented-out Float128 lines could use cross-references or explanations about why they're disabled

tkelman commented:
> This is also a good candidate for PkgEval before merging (since it changes the behavior of ccall)

I can do this, but it'll be best if the master baseline has fixed the ssl certificate issues that are currently causing problems.

vtjnash commented:
> Ptr{_<:T} is still really confusing in one of the ccall error messages, and the commented-out Float128 lines could use cross-references or explanations about why they're disabled

These are left alone because I'm trying not to change everything in one PR. Sure this changes a lot of related content. But that still doesn't mean it needs to change everything in sight.

vtjnash commented:
I did a PkgEval run. There's lots of unrelated failures (due to the mbedtls bug and the cache uuid bug in my build). And a lot more failures being picked up due to inadequate test coverage in those packages (this moves a compile error earlier). The set that might merit further research are below, although from a quick glance through, the failures there looked like the ones I expected should happen and will be easy for those authors to resolve.
https://gist.github.com/vtjnash/3a05a54fe793bb52734e2a74a9265682
Documenter
FLAC
GMT
HTTPClient
FactCheck (correctly detects that ccall parsing is no longer special)
LibCURL
LibExpat
NLopt
ParameterizedFunctions
RequestsCache
SFML

tkelman commented:
> due to the mbedtls bug and the cache uuid bug in my build

rebase the branch and those won't happen

vtjnash committed:
temporarily make invalid ccall syntax only a warning

a handful of packages seem to have unused functions that are triggering this error

vtjnash commented:
I reran PkgEval; it produced no interesting new results. The updated results are in the same gist above.

Merge?

tkelman commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -533,30 +539,40 @@ the field that will have the greatest size (potentially including padding).
+    b_a_2 = B.A[2]
 
-Arrays of unknown size are not supported.
+Arrays of unknown size (C99-complaint variable length structs specified by ``[]`` or ``[0]``) are not supported.
```
compliant

tkelman commented on src/ccall.cpp:
```diff
@@ -1667,163 +1721,235 @@ static jl_cgval_t emit_ccall(jl_value_t **args, size_t nargs, jl_codectx_t *ctx)
+#if JL_LLVM_VERSION >= 30800
+                instList.insertAfter(savespot->getIterator(), stacksave);
+#else
+                instList.insertAfter(savespot, stacksave);
```
wrong indent on these 2 lines?

tkelman commented on src/intrinsics.cpp:
```diff
@@ -355,104 +392,48 @@ static Value *emit_unbox(Type *to, const jl_cgval_t &x, jl_value_t *jt, Value *d
+    return mark_julia_type(r, true, (jl_value_t*)jl_any_type, ctx);
 }
 
 // put a bits type tag on some value (despite the name, this doesn't necessarily actually "box" the value however)
```
parenthetical no longer relevant?

tkelman commented on src/debuginfo.cpp:
```diff
@@ -822,7 +830,10 @@ static int lookup_pointer(DIContext *context, jl_frame_t **frames,
+        if (!frame->func_name ||
+                func_name.compare(0, 6, "jlsysw_") == 0 ||
+                func_name.compare(0, 6, "jlcall_") == 0 ||
+                func_name.compare(0, 7, "jlcapi_") == 0) {
```
why are `jlsysw_` and `jlcall_` comparing 6, but `jlcapi_` is comparing 7?

vtjnash committed:
avoid BoUpSLP vectorizer InsertElementInst bug in llvm-3.3

JeffBezanson commented on src/alloc.c:
```diff
@@ -267,8 +267,31 @@ JL_DLLEXPORT jl_value_t *jl_new_struct_uninit(jl_datatype_t *type)
 
+// type constructor -----------------------------------------------------------
+
+JL_DLLEXPORT jl_value_t *jl_new_type_constructor(jl_svec_t *p, jl_value_t *body)
```
Could you move this back? It will conflict with my branch.

JeffBezanson commented on base/inference.jl:
```diff
@@ -1201,19 +1186,19 @@ function abstract_eval_call(e::Expr, vtypes::VarTable, sv::InferenceState)
 
 function abstract_eval(e::ANY, vtypes::VarTable, sv::InferenceState)
-    if isa(e,QuoteNode)
+    if isa(e, QuoteNode)
```
Would be nice to take out these small whitespace changes.

JeffBezanson commented:
Also discussed: `box` is not a very accurate term; should probably be renamed `bitcast`.

JeffBezanson commented on src/julia.h:
```diff
@@ -255,10 +255,6 @@ typedef struct _jl_method_t {
-    // if there are intrinsic calls, sparams are probably required to compile successfully,
-    // and so unspecialized will be created for each linfo instead of using linfo->def->template
-    // 0 = no, 1 = yes, 2 = not yet known
-    uint8_t needs_sparam_vals_ducttape;
```
:tada: :100:

StefanKarpinski commented on src/codegen.cpp:
```diff
@@ -299,8 +298,15 @@ int32_t jl_jlcall_api(const void *function)
+    StringRef Name = F->getName();
+    if (Name.startswith("japi3_"))
+        return 3;
+    assert(Name.startswith("japi1_") ||
```
We should also probably call these jlapiN, no? Although I guess old ones might be an issue if we want to maintain C API compatibility.

vtjnash replied:
Julia ABI N would perhaps be the most accurate. But these are not externally visible names.

vtjnash committed:
remove unbox intrinsic

it was neutered by earlier commits, this just finishes deleting it

vtjnash committed:
remove box intrinsic

by adding type inference information to all of the intrinsics,
the box intrinsic is no longer required for this purpose

vtjnash committed:
remove jl_has_intrinsics bandaid

vtjnash committed:
rename box intrinsic to bitcast

this name better reflects what this intrinsic does,
which is to change the type-tag on some data,
without actually altering the data or its representation

vtjnash commented:
just to double-check that removing box/unbox didn't impact perf, @nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/18754#issuecomment-262691383) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/1f4ef86dbcc5d6ce375b5920d325be8e4af1c926/b9a8c42_vs_3daa183/report.md). cc @jrevels

vtjnash committed:
only acquire the method lock at most once during codegen

vtjnash committed:
make ccall fully statically compilable

and remove the ducttape!

vtjnash committed:
remove unnecessary calls to box intrinsic

tkelman commented on base/deprecated.jl:
```diff
@@ -1763,4 +1763,6 @@ export @test_approx_eq
 @deprecate(SharedArray{T}(filename::AbstractString, ::Type{T}, dims::NTuple, offset; kwargs...),
            SharedArray{T,length(dims)}(filename, dims, offset; kwargs...))
 
+is_intrinsic_expr(x::ANY) = false
```
missing `depwarn`

vtjnash commented:
@JeffBezanson now that jb/subtypes is merged, I think this should be ready to merge also

tkelman commented:
Not without a nanosoldier run, considering how long it's been since the last one.

tkelman commented on test/ccall.jl:
```diff
@@ -123,7 +126,11 @@ function test_struct1{Struct}(::Type{Struct})
 
     a2 = copy(a)
-    x = ccall((:test_1, libccalltest), Struct, (Struct, Float32), a2, b)
+    if Struct === Struct1
```
what's the reason for these splits now?

vtjnash replied:
The old version was taking advantage of the fact that we would usually compile a new version of the method for every input type. This PR removes that constraint from codegen so we can make it more robust.

yuyichao replied:
Are you saying that this will be error after the pr? That'll be a big regression.

vtjnash replied:
In practice, it's not an issue. When I ran PkgEval, and there weren't any packages that failed due to needing this.

yuyichao replied:
Is there an explicit rule about the use of type parameters in ccall?

vtjnash replied:
Yes. You can have them anywhere they don't affect the ABI

yuyichao replied:
It sounds ok for now as long as it doesn't break FunctionWrappers.jl **BEFORE** the alternative is implemented.

In the long term, I think we shouldn't have such restriction as long as it's possible to do efficiently **with** codegen and possible to implement without codegen (with libffi for example). As long as it is expressible, whether or not the construct requires compiler or not shouldn't be visible functionally (i.e. it can of course affect performance). Changing the ccall API to not allow this syntactically is of course different.

vtjnash replied:
FunctionWrappers.jl already doesn't like the current situation (https://github.com/yuyichao/FunctionWrappers.jl/blob/master/src/FunctionWrappers.jl#L77), so it won't be affected at all. The only change this makes is that now the argument type tuple ABI must be actually statically computable from the signature of the containing function (using `@generated` qualifies for this), instead of being required to be statically computable from "whatever we could constant propagate during codegen + some semi-arbitrary heuristics".

Don't worry, this is quite non-breaking in practice. Really it's just a simplification of the rule that the argument type tuple must be a constant literal. In the future, it might even allow us to relax that constraint (changing from literal -> evaluated at the toplevel during method definition).

vtjnash commented:
You do realize that you could just trigger @nanosoldier `runbenchmarks(ALL, vs=":master")` yourself?

tkelman commented:
right, though if this fails CI, re-running nanosoldier on an updated version takes longer

vtjnash commented:
Usually, but with the current backlog queue on CI, I think nanosoldier might finish first 😛

tkelman commented on src/ccall.cpp:
```diff
@@ -1133,49 +1179,90 @@ static jl_cgval_t emit_llvmcall(jl_value_t **args, size_t nargs, jl_codectx_t *c
+public:
+    std::vector<Type*> fargt; // vector of llvm output types (julia_struct_to_llvm) for arguments (vararg is the last item, if applicable)
+    std::vector<Type*> fargt_sig; // vector of ABI coercion types for call signature
+    std::vector<bool> fargt_isboxed; // vector of whether the llvm output types is boxed for each argument (vararg is the last item, if applicable)
```
"output type is boxed" or "output types are boxed"

vchuravy commented on src/ccall.cpp:
```diff
@@ -1648,166 +1728,242 @@ static jl_cgval_t emit_ccall(jl_value_t **args, size_t nargs, jl_codectx_t *ctx)
+        else {
+            assert(symarg.f_name != NULL);
+            llvmf = jl_Module->getOrInsertFunction(symarg.f_name, functype);
+            if (!isa<Function>(llvmf) || cast<Function>(llvmf)->getIntrinsicID() == Intrinsic::not_intrinsic)
```
As mention by Tim in https://github.com/JuliaLang/julia/pull/18754#issuecomment-250977056 it would be nice if we could turn off the second check. Maybe a second calling convention `llvmcall_unsafe`?

vchuravy commented on src/ccall.cpp:
```diff
@@ -1648,166 +1728,242 @@ static jl_cgval_t emit_ccall(jl_value_t **args, size_t nargs, jl_codectx_t *ctx)
+            jl_error("llvmcall doesn't support dynamic pointers");
+        }
+        else if (symarg.fptr != NULL) {
+            jl_error("llvmcall doesn't support static pointers");
```
Currently `llvmcall` has a second form that @Keno is using for `Cxx.jl` that takes a pointer to a `llvmf`. Not something that this needs to cover, but we should keep it in mind.

vtjnash replied:
yeah, I know. It's also seems to be fundamentally incompatible with the way we manage the rest of the code pipeline, so it's low on my priority lists of things that need to work. I've written up a replacement plan at https://github.com/vtjnash/julep/wiki/0002:-Enhanced-llvmcall (which I believe should also cover Tim's case below), but waiting for someone to actually care about the implementation. Maybe that person could be you :) :) ;) :)

vchuravy replied:
Right, I remember that.

 Are we planning on deprecating the current version of `llvmcall` for 1.0?
I have one example that I think still requires the current form and SIMD.jl might have issues as well.
Especially problematic is if you need to work with LLVM types that are not represented in Julia currently (e.g. `i1` and `i2`).

A long term alternative might be LLVM.jl allowing us to programmatically build up IR instead of relying on string representations.

vtjnash replied:
The above proposal should support a superset of the current implementation.

vchuravy reviewed (commented):
Nice! I found it easier to understand then before (especially intrinsics.cpp). Looks to me like an overall simplification and would be great to have for v0.6

tkelman commented on doc/src/manual/calling-c-and-fortran-code.md:
```diff
@@ -822,6 +822,20 @@ err = ccall(:gethostname, stdcall, Int32, (Ptr{UInt8}, UInt32), hn, length(hn))
+There is one additional special calling convention `llvmcall`,
+which allows inserting calls to LLVM intrinsics directly.
+This can be especially useful when targeting unusual platforms such as GPGPUs.
+For example, for [CUDA](http://llvm.org/docs/NVPTXUsage.html), we need to be able to read the thread index::
```
the double-colon might be a leftover rst-ism?

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/18754#issuecomment-272979848) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/033ed8d16c5c23a442545e8ae68773f66f9f37af/02f17c8_vs_a1f232e/report.md). cc @jrevels

tkelman commented:
I was just about to say I got the same assertion `jl_typeis(a, jl_array_any_type), file c:/projects/julia/src/array.c, line 1042` to fail, same as on mac travis (while running bootstrap instead of a test, but different build configurations).

tkelman commented on src/intrinsics.cpp:
```diff
@@ -1040,156 +901,190 @@ static jl_cgval_t emit_intrinsic(intrinsic f, jl_value_t **args, size_t nargs,
-        assert(z->getType() == y->getType());
-        Value *fmaintr = Intrinsic::getDeclaration(jl_Module, Intrinsic::fma,
-                                   ArrayRef<Type*>(x->getType()));
+          assert(y->getType() == x->getType());
```
indentation is off

vtjnash commented:
@vchuravy thanks! That's been my goal :) I think that it's much easier to express code-generation as a set of optimizations over a static implementation, rather than hoping that you can express all dynamic code-generation paths (even in the presence of a inference algorithm that can't always give you those sorts of guarantees). I know this caused you a bit of pain, since I wouldn't let you layer more optimizations on top that didn't have a static implementation.

It looks like I was missing a `jl_gc_wb` on that line.

tkelman commented:
If this is at all breaking, it should have a news mention and possibly doc updates

JeffBezanson commented on base/serialize.jl:
```diff
@@ -39,7 +39,7 @@ const TAGS = Any[
     :p, :q, :r, :s, :t, :u, :v, :w, :x, :y, :z,
     :add_int, :sub_int, :mul_int, :add_float, :sub_float,
-    :mul_float, :unbox, :box,
+    :mul_float, :bitcast, :box,
```
rm `box`?

vtjnash replied:
I'm assuming someone will change it to `reservedN` at some point or reuse it, but that keeping it in place keeps the indexing the same for `ser_version`.

JeffBezanson replied:
Doesn't changing `unbox` to `bitcast` also break the serialization format though?

vtjnash replied:
no idea

vtjnash committed:
add docs and news for ccall change

JeffBezanson commented on base/inference.jl:
```diff
@@ -340,7 +342,95 @@ function instanceof_tfunc(t::ANY)
     return Any
 end
-add_tfunc(box, 2, 2, (t,v)->instanceof_tfunc(t))
+box_tfunc(t::ANY, x::ANY) = instanceof_tfunc(t)
```
`bitcast_tfunc`?

tkelman commented on doc/src/manual/calling-c-and-fortran-code.md:
```diff
@@ -260,29 +269,37 @@ same:
   * `Ref{T}`
 
-    Behaves like a `Ptr{T}` that owns its memory.
+    Behaves like a ``Ptr{T}`` that can manage its memory via the Julia GC.
```
single backticks for markdown

tkelman commented on doc/src/manual/calling-c-and-fortran-code.md:
````diff
@@ -420,20 +437,48 @@ struct B {
+b_a_2 = B.A[2]
+```
+
+Arrays of unknown size (C99-compliant variable length structs specified by ``[]`` or ``[0]``) are not directly supported.
````
single backticks for markdown

tkelman commented on doc/src/manual/calling-c-and-fortran-code.md:
````diff
@@ -420,20 +437,48 @@ struct B {
+```
+
+Arrays of unknown size (C99-compliant variable length structs specified by ``[]`` or ``[0]``) are not directly supported.
+Often the best way to deal with these is to deal with the bytes offsets directly.
````
"byte offsets" ?

tkelman commented on doc/src/manual/calling-c-and-fortran-code.md:
````diff
@@ -420,20 +437,48 @@ struct B {
+In Julia, we can access the parts independently to make a copy of that string:
+
 ```
+str = from-c::Ptr{Void}
````
`from_c` ? this is slightly confusing

JeffBezanson commented on base/inference.jl:
```diff
@@ -1431,17 +1511,47 @@ function abstract_eval(e::ANY, vtypes::VarTable, sv::InferenceState)
+            spsig = sv.linfo.def.sig
+            if isa(spsig, UnionAll)
+                env = data_pointer_from_objref(sv.linfo.sparam_vals) + sizeof(Ptr{Void})
+                rt = ccall(:jl_instantiate_type_in_env, Any, (Any, Any, Ptr{Any}), e.args[2], spsig, env)
```
Is it worth checking for free type vars before doing this?

vtjnash replied:
Is it? I assumed that apply_type was efficient at handling that case.

JeffBezanson commented on base/inference.jl:
```diff
@@ -2650,16 +2760,30 @@ function substitute!(e::ANY, na, argexprs, spvals, offset)
+        elseif head === :foreigncall
+            for i = 1:length(e.args)
+                if i == 2
+                    e.args[2] = ccall(:jl_instantiate_type_in_env, Any, (Any, Any, Ptr{Any}), e.args[2], spsig, spvals)
```
Does this mean these types only get instantiated when the code is inlined somewhere?

vtjnash replied:
Yes

JeffBezanson commented on base/int.jl:
```diff
@@ -427,12 +423,12 @@ widen(::Type{UInt64}) = UInt128
-widemul(x::Number,y::Bool) = x*y
+widemul(x::Bool,y::Bool) = x * y
+widemul(x::Bool,y::Number) = x * y
+widemul(x::Number,y::Bool) = x * y
```
I don't know if it's worth changing at this point, but the large number of whitespace-only changes in here is annoying.
